### DISCUSSION TITLE
Dirty attributes on elements

### DISCUSSION THREAD

**Discussion by jelmer-scoremedia** [craftcms/cms#14593] [General]
Setting field values on element performs the following logic:

```
public function setFieldValue(string $fieldHandle, mixed $value): void
    {
        $behavior = $this->getBehavior('customFields');
        $behavior->$fieldHandle = $value;

        // Don't assume that $value has been normalized
        unset($this->_normalizedFieldValues[$fieldHandle]);

        // If the element is fully initialized, mark the value as dirty
        if ($this->_initialized) {
            $this->_dirtyFields[$fieldHandle] = true;
        }

        // If the field value was previously eager-loaded, undo that
        unset($this->_eagerLoadedElements[$fieldHandle]);
        unset($this->_eagerLoadedElementCounts[$fieldHandle]);
    }
```

_According to documentation:_
You can tell whether **field content has changed** by calling [isFieldDirty()(opens new window)](https://docs.craftcms.com/api/v4/craft-base-elementinterface.html#method-isfielddirty) on the element.

But actually is **not true**, if you set the same value the property will be marked as dirty **while content has not been changed.**

```
$product->field
// foo

$product->isFieldDirty('field');
// false

$product->setFieldValue('field', 'foo');

$product->isFieldDirty('field');
// true
```

**Comment by brandonkelly** [comment]
Correct. It’s not super smart right now, because it would be pretty expensive for fields like Matrix to determine whether something within the value is actually different than the current value.

It’s not really an issue when editing elements in the control panel though, because the edit form will only post field values that have changed. So `setFieldValue()` only ends up getting called when something changes.
